pysd/py_backend/external.py

```diff
@@ -0,0 +1,962 @@
+"""
+These classes are a collection of the needed tools to read external data.
+The External type objects created by these classes are initialized before
+the Stateful objects by functions.Model.initialize.
+"""
+
+import re
+import os
+import warnings
+import pandas as pd
+import numpy as np
+import xarray as xr
+from . import utils
+
+try:
+    # Optional dependency as openpyxl requires python 3.6 or greater
+    # Used for reading data giving cell range names
+    from openpyxl import load_workbook
+
+except Exception:
+    warnings.warn(
+      "Not able to import openpyxl.\n"
+      + "You will not be able to read Excel data by cell range names.\n"
+      + " You can read Excels in the usual way as long as the reading "
+      + "information is by cell name (such as 'A1') and by row number "
+      + "or column letter in the case of GET DATA or GET LOOKUPS.\n\n"
+      )
+
+
+class Excels():
+    """
+    Class to save the read Excel files and thus avoid double reading
+    """
+    _Excels, _Excels_opyxl = {}, {}
+
+    @classmethod
+    def read(cls, file_name):
+        """
+        Read the Excel file or return the previously read one
+        """
+        if file_name in cls._Excels:
+            return cls._Excels[file_name]
+        else:
+            excel = pd.ExcelFile(file_name)
+            cls._Excels[file_name] = excel
+            return excel
+
+    @classmethod
+    def read_opyxl(cls, file_name):
+        """
+        Read the Excel file using OpenPyXL or return the previously read one
+        """
+        if file_name in cls._Excels_opyxl:
+            return cls._Excels_opyxl[file_name]
+        else:
+            excel = load_workbook(file_name, read_only=True)
+            cls._Excels_opyxl[file_name] = excel
+            return excel
+
+    @classmethod
+    def clean(cls):
+        """
+        Clean the dictionary of read files
+        """
+        cls._Excels, cls._Excels_opyxl = {}, {}
+
+
+class External(object):
+    """
+    Main class of external objects
+    """
+
+    def __init__(self, py_name):
+        self.py_name = py_name
+
+    def __str__(self):
+        return self.py_name
+
+    def _get_data_from_file(self, rows, cols):
+        """
+        Function to read data from excel file using rows and columns
+
+        Parameters
+        ----------
+        rows: int, list or array
+            if int, then will start reading the table for the given row number
+            if list, then will read between first and second row of the lis
```

Comment:
not sure I understand this - is it asking for a (first, last) tuple? Same for cols. If we haven't already, can you add an example to the documentation showing how to use these parameters?

Reply:
Can you request just two rows, or will it interpret this as a first/last?

Reply:
You can request two rows if you use an array. Is a little bit confusing but this function is internally working and all the cases work well So nothing to worry about.
In 1.0.0 we can move all the excel reading to openpyxl, to remove pandas dependency in this file. Then we will need to rebuild these functions and we can do it in a more convenient way.

Reply:
I guess what I'm worried about is communicating to the user. I'm good with reworking it for 1.0.0, just so long as the API we settle on is explicit. Can you open an issue to describe the different ways we could handle that case?

Reply:
Let's continue it in #228